File: node/src/ArduinoManager.js
const fs = require('fs')
const ArduinoInterface = require('./ArduinoInterface')

module.exports = class ArduinoManager {

  constructor() {
    this.instances = []
  }

  /**
   * We will look every serial port
   * We will filter by vendor id
   * We will send each a command to known the label ('stepper' or 'encoder')
   */
  bindArduino() {
    return new Promise(async (resolve) => {
      let dir = fs.readdirSync('/dev').filter(dir => dir.indexOf('ttyUSB') > -1)
      let promises = []
      for (var i = 0; i < dir.length; i++) {
        let arduino = new ArduinoInterface('/dev/' + dir[i])
        this.instances.push(arduino)
        promises.push(arduino.init())
      }

      await Promise.all(promises)

      this.instances.forEach(instance => console.log(`> ARDUINO: detected ${instance.getLabel()} on port ${instance.getPath()}`))

      resolve()
    })
  }

  /**
   * Will return the Arduino interface for the encoder arduino
   */
  getEncoderArduino() {
    return this.instances.filter(arduino => arduino.getLabel() === 'ENCODER')[0]
  }

  /**
   * Will return the Arduino interface for the stepper arduino
   */
  getStepperArduino() {
    return this.instances.filter(arduino => arduino.getLabel() === 'STEPPER')[0]
  }

  getInstanceInventory() {
    return this.instances.map(a => {
      let port = a.getPort()
      return {
        label: a.getLabel(),
        path: a.getPath(),
        instance: (port !== null && port !== undefined && typeof port === 'object') ? port.constructor.name : port
      }
    })
  }

}

File: python/src/ArduinoManager.py
import os
from .Arduino import Arduino

class ArduinoManager:
  def __init__(self, container):
    self.instances = []
    self.container = container
    self.logger = self.container.get('logger').get('ArduinoManager')
  
  def identify(self):
    # get the list of all the ttyUSBX devices in /dev
    # for each of them create a instance
    devices = os.popen('ls /dev | grep "ttyUSB"').read().split("\n")[0:-1]

    #results = Parallel(n_jobs=-1)(delayed(self.identifySingle)(d) for d in devices)
    results = []
    for d in devices:
      results.append(self.identifySingle(d))

    for device in results:
      name = device.getName()
      if 'STEPPER' in name:
        self.container.set('arduinoStepper', device)
        device.setName('STEPPER')
        self.logger.info('Found STEPPER')
      if 'SWITCHES' in name:
        self.container.set('arduinoSwitches', device)
        device.setName('SWITCHES')
        self.logger.info('Found SWITCHES')
    return results

  def identifySingle(self, device):
    instance = Arduino('/dev/' + device)
    instance.init()
    instance.identify()
    return instance
